### PR TITLE
V2

### PR DESCRIPTION
**Pull request by ignaciojonas**

### CODE REVIEW

File: lib/auth0/api/v2/connections.rb [hunk 1]
```diff
@@ -29,14 +51,37 @@ def connection(connection_id, fields: nil, include_fields: true)
...
+
+        # Updates a connection. Updates the fields specified in the body parameter.
+        # @see https://auth0.com/docs/api/v2#!/Connections/patch_connections_by_id
+        # @param connection_id [string] The id of the connection to delete
+        # @param body [hash] The Hash options used to update the conecctions's properties.
+        #
+        # @return  [json] Returns the updated connection.
         def update_connection(connection_id, body)
-          fail Auth0::MissingConnectionId, 'you must specify a connection id' if connection_id.to_s.empty?
+          fail Auth0::InvalidParameter, 'Must supply a valid connection id' if connection_id.to_s.empty?
```
Comment (leosoubeste) [inline lib/auth0/api/v2/connections.rb:84]: This format does not match on all validations for the rest of the endpoints. What form should we use: "you must..." or "Must..."?

File: spec/integration/lib/auth0/api/v2/api_jobs_spec.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      let(:connection_id) do
+        client.connections
+          .find do |connection|
+            connection['name'].include?(Auth0::Api::AuthenticationEndpoints::UP_AUTH)
+          end['id']
+      end
+      let(:imported_users) { client.import_users(users_file, connection_id) }
+      it do
+        expect(imported_users).to include(
+          'connection' => 'Username-Password-Authentication',
```
Comment (leosoubeste) [inline spec/integration/lib/auth0/api/v2/api_jobs_spec.rb]: Should we use constant value here?

Comment (ignaciojonas) [inline spec/integration/lib/auth0/api/v2/api_jobs_spec.rb]: Yes, we have to use the constant here. Good catch!

File: lib/auth0/api/v2/connections.rb (hunk 1 above)
Comment (ignaciojonas) [inline lib/auth0/api/v2/connections.rb:84]: Most of the errors are 'Must....;, so we'll keep this one, but we should change it in clients.rb

File: lib/auth0/api/v2/clients.rb [hunk 3]
```diff
@@ -1,37 +1,71 @@
...
         alias_method :get_clients, :clients
 
-        # https://auth0.com/docs/apiv2#!/clients/post_clients
+        # Creates a new client application.
+        # @see https://auth0.com/docs/api/v2#!/clients/post_clients
+        # @param name [string] The name of the client. Must contain at least one character. Does not allow '<' or '>'
+        # @param options [hash] The Hash options used to define the client's properties.
+        # @return  [json] Returns the created client application.
         def create_client(name, options = {})
+          fail Auth0::MissingParameter, 'you must specify a valid client name' if name.to_s.empty?
```
Comment (ignaciojonas) [inline lib/auth0/api/v2/clients.rb]: This should be 'Must...'  and all the message in this class too.

Comment (ignaciojonas) [inline lib/auth0/api/v2/clients.rb]: BTW, we should also update the tests.

File: lib/auth0/api/v2/connections.rb (hunk 1 above)
Comment (leosoubeste) [inline lib/auth0/api/v2/connections.rb:84]: Changed

File: lib/auth0/api/v2/users.rb [hunk 4]
```diff
@@ -30,39 +50,107 @@ def create_user(name, options = {})
...
+        # property too.
+        # If your are updating email or phone_number you need to specify the connection and the client_id properties.
+        # @see https://auth0.com/docs/api/v2#!/Users/patch_users_by_id
+        # @param user_id [string] The user_id of the user to update.
+        # @param options [hash] The optional parametes to update
+        #
+        # @return [json] the updated user
         def patch_user(user_id, options)
+          fail Auth0::MissingUserId, 'Must supply a valid user_id' if user_id.to_s.empty?
+          fail Auth0::InvalidParameter, 'Must supply a valid body' if options.to_s.empty?
```
Comment (falvarisqueta) [inline lib/auth0/api/v2/users.rb]: Message says Must supply body and parameter is options. Is this message set by requirement?

File: lib/auth0/api/v2/users.rb [hunk 5]
```diff
@@ -30,39 +50,107 @@ def create_user(name, options = {})
...
+        def unlink_users_account(user_id, provider, secondary_user_id)
+          fail Auth0::MissingUserId, 'Must supply a valid user_id' if user_id.to_s.empty?
+          fail Auth0::MissingUserId, 'Must supply a valid secondary user_id' if secondary_user_id.to_s.empty?
+          fail Auth0::InvalidParameter, 'Must supply a valid provider' if provider.to_s.empty?
+          path = "/api/v2/users/#{user_id}/identities/#{provider}/#{secondary_user_id}"
+          delete(path)
+        end
       end
     end
   end
```
Comment (falvarisqueta) [inline lib/auth0/api/v2/users.rb:160]: api/v2/users is repeated in the file, should we add a base_path?

File: spec/integration/lib/auth0/api/v2/api_connections_spec.rb [hunk 6]
```diff
@@ -53,16 +53,45 @@
 
   describe '.delete_connection' do
     it { expect { client.delete_connection connection['id'] }.to_not raise_error }
-    it { expect { client.delete_connection '' }.to raise_error(Auth0::MissingConnectionId) }
```
Comment (falvarisqueta) [inline spec/integration/lib/auth0/api/v2/api_connections_spec.rb:56]: what's the reason to remove this expectation?

File: spec/integration/lib/auth0/api/v2/api_connections_spec.rb [hunk 7]
```diff
@@ -53,16 +53,45 @@
...
+  describe '.delete_connection_user' do
+    let(:username) { Faker::Internet.user_name }
+    let(:email) { "#{entity_suffix}#{Faker::Internet.safe_email(username)}" }
+    let(:password) { Faker::Internet.password }
+    let!(:user_to_delete) do
+      client.create_user(username,  'email' => email,
+                                    'password' => password,
+                                    'email_verified' => false,
+                                    'connection' => Auth0::Api::AuthenticationEndpoints::UP_AUTH,
+                                    'app_metadata' => {})
```
Comment (falvarisqueta) [inline spec/integration/lib/auth0/api/v2/api_connections_spec.rb]: client.create_connection (strategy: strategy) and client.create_user(email => email) have different formats, is this on purpose?

File: lib/auth0/api/v2/users.rb (hunk 4 above)
Comment (ignaciojonas) [inline lib/auth0/api/v2/users.rb]: Good Catch! Changed

File: spec/integration/lib/auth0/api/v2/api_connections_spec.rb (hunk 6 above)
Comment (ignaciojonas) [inline spec/integration/lib/auth0/api/v2/api_connections_spec.rb:56]: We are covering this in our unit tests.

**Comment by ignaciojonas** [comment]
All feedback applied. Merging.

**Pull request merged**